libyul/backends/wasm/EVMToEWasmTranslator.cpp

```diff
@@ -0,0 +1,370 @@
+#include <libyul/optimiser/Disambiguator.h>
+#include <libyul/optimiser/NameDisplacer.h>
+
+#include <libyul/AsmParser.h>
+#include <libyul/AsmAnalysis.h>
+#include <libyul/AsmAnalysisInfo.h>
+#include <libyul/Object.h>
+
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/Scanner.h>
+#include <liblangutil/SourceReferenceFormatter.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+using namespace langutil;
+
+namespace
+{
+static string const polyfill{R"({
+function or_bool(a, b, c, d) -> r {
+	r := i64.ne(0, i64.or(i64.or(a, b), i64.or(c, d)))
+}
+function add_carry(x, y, c) -> r, r_c {
+	let t := i64.add(x, y)
+	r := i64.add(t, c)
+	r_c := i64.or(
+		i64.lt_u(t, x),
+		i64.lt_u(r, t)
+	)
+}
+function add(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+	let carry
+	r4, carry := add_carry(x4, y4, 0)
+	r3, carry := add_carry(x3, y3, carry)
+	r2, carry := add_carry(x2, y2, carry)
+	r1, carry := add_carry(x1, y1, carry)
+}
+function bit_negate(x) -> y {
+	y := i64.xor(x, 0xffffffffffffffff)
+}
+function sub(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+	// x - y = x + (~y + 1)
+	let carry
+	r4, carry := add_carry(x4, bit_negate(y4), 1)
+	r3, carry := add_carry(x3, bit_negate(y3), i64.add(carry, 1))
+	r2, carry := add_carry(x2, bit_negate(y2), i64.add(carry, 1))
+	r1, carry := add_carry(x1, bit_negate(y1), i64.add(carry, 1))
+}
+function split(x) -> hi, lo {
+	hi := i64.shr_u(x, 32)
+	lo := i64.and(x, 0xffffffff)
+}
+// Multiplies two 64 bit values resulting in a 128 bit
+// value split into two 64 bit values.
+function mul_64x64_128(x, y) -> hi, lo {
+	let xh, xl := split(x)
+	let yh, yl := split(y)
+
+	let t0 := i64.mul(xl, yl)
+	let t1 := i64.mul(xh, yl)
+	let t2 := i64.mul(xl, yh)
+	let t3 := i64.mul(xh, yh)
+
+	let t0h, t0l := split(t0)
+	let u1 := i64.add(t1, t0h)
+	let u1h, u1l := split(u1)
+	let u2 := i64.add(t2, u1l)
+
+	lo := i64.or(i64.shl(u2, 32), t0l)
+	hi := i64.add(t3, i64.add(i64.shr_u(u2, 32), u1h))
+}
+function add3_carry(a, b, c, carr) -> x, carry {
+	x, carry := add_carry(a, b, carr)
+	x, carry := add_carry(x, c, carry)
+}
+// Multiplies two 128 bit values resulting in a 256 bit
+// value split into four 64 bit values.
+function mul_128x128_256(x1, x2, y1, y2) -> r1, r2, r3, r4 {
+	let ah, al := mul_64x64_128(x1, y1)
+	let     bh, bl := mul_64x64_128(x1, y2)
+	let     ch, cl := mul_64x64_128(x2, y1)
+	let         dh, dl := mul_64x64_128(x2, y2)
+	let carry
+	r4 := dl
+	r3, carry := add3_carry(bl, cl, dh, 0)
+	r2, carry := add3_carry(al, bh, ch, carry)
+	r1 := i64.add(ah, carry)
+}
+function mul(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
+}
+function div(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function mod(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function smod(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+function exp(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {}
+
+function byte(x1, x2, x3, x4, y1, y2, y3, y4) -> r1, r2, r3, r4 {
```

Comment:
I don't understand what this function is doing, perhaps add comments to explain?

Reply:
It's one of the EVM opcodes.

Reply:
Oh that makes sense. https://ethervm.io/#1A looks like it should count from least significant byte?

Reply:
ethervm.io seems to be wrong. Citing the yellow paper: "For the Nth byte, we count from the left (i.e.  N=0 would be the most significant in big endian)"

Reply:
Oh... Wow, ok. I'm writing an email to tell them about this.